### ISSUE TITLE
Resolve BUILDKITE_COMMIT after checkout?

### ISSUE THREAD

**Issue by sj26** [buildkite/agent#1188]
When `$BUILDKITE_COMMIT` is `HEAD`, would it make sense to update `$BUILDKITE_COMMIT` to the resolved commit after checkout? We already do this for pipeline upload jobs:

https://github.com/buildkite/agent/pull/871

but not for pipelines which don't include a pipeline upload step. This means a pipeline with a default build pipeline configured from the dashboard and a scheduled build at `master` and `HEAD` will have a `$BUILDKITE_COMMIT` value of `HEAD` even after checkout, and may fail if the first job includes logic which presumes that `$BUILDKITE_COMMIT` holds an actual commit.

Perhaps this could be introduced as an agent bootstrap option `BUILDKITE_AGENT_RESOLVE_COMMIT_AFTER_CHECKOUT` or something which will default to true after v4?

**Comment by toolmantim** [comment]
Could you share more about why were you worried about making this the default behavior immediately? At a glance, it seems more like a bug that needs fixing, vs a breaking change?

**Comment by sj26** [comment]
True, the value is sent back to the buildkite api and becomes the BUILDKITE_COMMIT for all future jobs, maybe committing (lol) the change locally as well makes sense and should be considered a bug. I was just worried about plugins or pipelines which might be relying on this behaviour somehow, but I can't think of a good reason, and haven't seen an actual case, so I'm probably being overly cautious. 😅

**Comment by keithduncan** [comment]
I just ran in to this today 😄

I knew that `BUILDKITE_COMMIT` was a commit sha in some of my pipelines, but wasn’t sure why my newer ones didn’t have this. Searching I found https://github.com/buildkite/agent/pull/871 and realised that since I’ve switched to using the YAML steps editor instead of pipeline upload step my pipelines don’t get a resolved sha ☹️

**Comment by ailisp** [comment]
I was using BUILDKITE_COMMIT to tag build docker images xxx:$BUILDKITE_BRANCH-$BUILDKITE_COMMIT, but this results image names to be xxx:master-HEAD.  To work around this i have to do:
```
      if [[ ${BUILDKITE_COMMIT} == "HEAD" ]]; then
        export BUILDKITE_COMMIT=$(git rev-parse HEAD)
      fi
```
it would be great to have both `BUILDKITE_COMMIT` and `BUILDKITE_COMMIT_HASH` (or `BUILDKITE_COMMIT_RESOLVED`, And I'd also suggest adding `BUILDKITE_COMMIT_SHORT` which is the short version of commit hash

**Comment by lox** [comment]
Unfortunately in the YAML steps editor, we don't actually have access to the git repo, so we can't resolve references like `HEAD` until we have code checked out. 

@ailisp I'm interested to understand why https://github.com/buildkite/agent/pull/871 hasn't fixed your problem.

**Comment by lox** [comment]
Oh, because it doesn't use a pipeline upload I am assuming.

**Comment by ailisp** [comment]
Yes I'm using the YAML steps editor, that you explained HEAD isn't resolved.

**Comment by chloeruka** [comment]
Closing this issue as we've made an experimental flag in above merged PR #1256 that should address this. We'll promote this for the next major agent version.
